Libplanet.Net/Consensus/Context.Mutate.cs

```diff
@@ -266,7 +266,24 @@ private void ProcessHeightOrRoundUponRules(ConsensusMsg message)
                         block4.Index,
                         block4.Hash,
                         ToString());
-                    _blockChain.Append(block4, GetBlockCommit());
+
+                    bool found = _blockValidationCache.TryGet(
+                        (block4.Hash, block4.StateRootHash), out var cached);
```

Comment:
I think new instances of anon pairs will have different `GetHashCode()` result. I don't think this would work without a new `struct`. Besides, I think `BlockHash` would suffice as a key. 🙄

Reply:
Also instead of trying to grab a cache here again, I think `bool IsValid(Block<T>, out IReadOnlyList<ActionEvaluation>)` might work better. 😗

Reply:
Oh, good point. It does not pair the block with actions but it only increases the input value for the cyclic hash.

_After some searches, found that those are boxed into `ValueTuple` and the hash code is calculated from the `GetHashCode()` of items. I'm not sure but probably `ValueTuple` will be compared with the hash code. (Because in this case, `LRUCache` uses `Dictionary.ContainsKey()` internally and it uses the hash code if `IEqualityComparer<T>` is not defined.)_

https://github.com/dotnet/roslyn/blob/f5a0fc0b9ddfdd58ee2609b36076059c6a0069e1/src/Compilers/Test/Resources/Core/NetFX/ValueTuple/ValueTuple.cs#L637-L641

Reply:
I guess I was mistaken. 🙄

Reply:
But, your statement seems to be right, [[stackoverflow link goes here]](https://stackoverflow.com/questions/2240231/how-well-does-net-dictionary-resolve-collisions). It seems .net `Dictionary` is handling collision internally. If we want this to be properly handled, `Equal()` should be implemented in `struct` or so.